compiler/pipes/convert-sprintf-calls.cpp

```diff
@@ -0,0 +1,179 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2021 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#include "compiler/pipes/convert-sprintf-calls.h"
+
+#include <compiler/gentree.h>
+
+struct FormatCallInfo {
+  FormatCallInfo() : args(VertexRange(Vertex::iterator{}, Vertex::iterator{})) {}
+  // stores a variable that contains all the call arguments,
+  // if all arguments are constant
+  VertexAdaptor<op_var> var;
+  // stores an array that contains all the call arguments,
+  // if at least one argument is a function call or a variable
+  VertexAdaptor<op_array> array;
+  // flag for the presence of a value in format_args_var
+  bool is_var{false};
+  // stores a set of vertices of the call arguments from var or array
+  VertexRange args;
+};
+
+struct FormatPart {
+  enum class FormatSpecType {
+    Decimal,
+    String
+  };
+
+  std::string value;
+  FormatSpecType specifier;
+
+  bool is_specifier() const {
+    return value.empty();
+  }
+};
+
+std::vector<FormatPart> try_parse_format_string(const std::string &format) {
+  std::vector<FormatPart> parts;
+  std::string last_value;
+  bool find_percent = false;
+
+  for (const auto &symbol : format) {
+    if (find_percent) {
+      if (symbol == 'd') {
+        parts.push_back(FormatPart{"", FormatPart::FormatSpecType::Decimal});
```

Comment:
BTW, this can be rewritten as
```
parts.emplace_back("", FormatPart::FormatSpecType::Decimal);
```

Reply:
> BTW, this can be rewritten as
> 
> ```
> parts.emplace_back("", FormatPart::FormatSpecType::Decimal);
> ```

Since `FormatPart` has no constructors, it seems that it won't be possible to rewrite this way. Perhaps replacing `push_back` with `emplace_back` would be okay.

Reply:
Fixed.